### DISCUSSION TITLE
Circular dependencies with state

### DISCUSSION THREAD

**Discussion by prabirshrestha** [Totodore/socketioxide#236] [Q&A] [answered]
I have a scenario where I'm hiring circular dependencie issue and wondering what is the best way to solve this.

I have an `AuthService` which allows me to validate jwt token as well as signin and signout. When I signin or signout I want to send send a notification, but then I want to also validate the jwt token. 

```rust
let (layer, io) = SocketIo::builder().with_state(AuthService::new()).build_layer();
```

As you can see I can't pass `io` to `AuthService` as it hasn't been created. While I could easily break jwt validation and signin and singout and remove the dependencies, I'm facing similar issues with other states.

Wondering if there is some way to set the state once `io` has been created this way I can use `AuthService` state during `on_connect` and `signin` and `signout` can also use `io` and emit messages.

```rust
let (layer, mut io) = SocketIo::builder().build_layer();
io.set_state(AuthService::new(io));
```

**Comment by tausifcreates** [comment]
What does the `AuthService` look like? I would set the `io` in a global variable after initialization, then I can use it in any route.

**Comment by Totodore** [comment]
There is no way to set state on `SocketIo` once it has been initialized for the moment :/. So you have to have a global variable or maybe set the io handle on your auth service after the initialization.

**Reply by Totodore** [reply]
You could also pass the `io` struct only when you need to perform the action with your service rather than holding a ref to it. But it depends on your `AuthService` implementation.

**Comment by prabirshrestha** [comment]
My `AuthService` looks something like this.

```rust
#[async_trait]
pub trait AuthService: Sync + Send {
    async fn decode_token(&self, token: &str) -> AppResult<DecodedToken>;
    async fn verify_password(&self, username: &str, password: &Secret<String>) -> AppResult<()>;
    async fn change_password_by_username(
        &self,
        username: &str,
        change_password_request: &ChangePasswordRequest,
    ) -> AppResult<()>;

    async fn update_user(&self, id: &str, update_user_request: UpdateUserRequest) -> AppResult<()>;

    /// ..
}
```

In `change_password_by_username` and `update_user` I would like to send notification when the function has been called.

```rust
async fn on_connect(socket: SocketRef, auth_service: State<Arc<AuthService>>) {
    let headers = &socket.req_parts().headers;
    let jwt = get_jwt_token_from_header(&headers).unwrap_or_default();
    
    let decoded_token = auth_service.decode_token(jwt).await;
    // ...
}
```

As you can see `auth_service` is used to decode token in `on_connect` but also requires access to `io`.

I would like to avoid global variables. I was sort of hoping if we could have some way to set a state later in io, such that we state can be of type `Option<Arc<AuthService>>`.

**Reply by tausifcreates** [reply]
Hm, can you initialize auth_service with `io`  struct passed down to it, after io is created? That way you will not need a global variable.

From what I see, the `AuthService` is a collection of auth utils, so it might not need to be a state?

It would be nice if the state could be set afterwards, but for now, you might want to consider the other approaches.

**Reply by Totodore** [reply, marked as answer]
You could also have a fn `set_io` that would set it on the AuthService but it would require to put a Rwlock.

But I think the best way is to just have functions that takes a SocketIo ref as argument each time it is needed.

**Reply by prabirshrestha** [reply]
I went with passing io as arguments to avoid RwLock.


```rust
let services = Arc::new(...); // this contains a bunch of services
let (io_layer, io) =  SocketIo::builder().build_layer();
let services = Arc::new(Services::new(io.clone()).await?);
io.ns("/", {
  let services = services.clone();
  |socket: SocketRef| async move { on_connection(socket, services).await }
});
```

**Comment by neckaros** [comment]
Thank for the impl exemple. I ended up doing the same thing.

@Totodore would it be possible to have an extractor that give us the SocketIo ?

**Reply by Totodore** [reply]
It requires some changes in the codebase architecture but it is planned.

**Reply by neckaros** [reply]
Great and thanks for the great work don on this library!
